### PR TITLE
feat: lights 

### PR DESCRIPTION
**Pull request by JoshuaVHB**
added points, dir and spots
also some changes to imgui

will probably do billboardss after, and then editor work

### CODE REVIEW

File: PyriteEditor/src/CornellBoxScene.h [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+            m_camController.processUserInputs(delta);
+            pcameraBuffer->setData(CameraBuffer::data_t{
+                .mvp = m_camera.getViewProjectionMatrix(),
+                .pos = m_camera.getPosition()
+                });
+            pinvCameBuffer->setData(InverseCameraBuffer::data_t{
+                .inverseViewProj = m_camera.getViewProjectionMatrix().Invert(),
+                .inverseProj = m_camera.getProjectionMatrix().Invert(),
+                .Proj = m_camera.getProjectionMatrix()
+                });
```
Comment (Akahara) [inline PyriteEditor/src/CornellBoxScene.h:139]: Shouldn't `setData` calls be in `render()`? To differenciate engine/graphic tasks

File: PyriteEditor/src/CornellBoxScene.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            std::copy_n(m_lights.ConvertCollectionToHLSL().begin(), 16, std::begin(light_data.lights));
+            pLightBuffer->setData(light_data);
+        }
+
+        /// ------------------------------------------------------------------------------------------------------------------------------- ///
+        
+        void render() override
+
+        {
+            pyr::Engine::d3dcontext().IASetPrimitiveTopology(D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
```
Comment (Akahara) [inline PyriteEditor/src/CornellBoxScene.h:151]: Could become a part of the RenderProfiles at some point

File: PyriteEditor/src/editor/bridges/Lights/pf_Light.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		std::string name;
+		Type type;
+	};
+
+	struct pf_LightsCollection
+	{
+		pf_LightsCollection() {}
+		pf_LightsCollection(pyr::LightsCollections* source)
+		{
+			assert(source);
```
Comment (Akahara) [inline PyriteEditor/src/editor/bridges/Lights/pf_Light.h:27]: Could be a `PYR_ASSERT`

File: PyriteEditor/src/editor/views/Lights/widget_lights.h [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+			mutable pf_LightsCollection LightsCollectionView;
+
+			virtual void display() override
+			{
+				if (!LightsCollectionView.sourceCollection) return;
+				LightsCollectionView.UpdateEditorFormat();
+
+				static pyr::Texture AddIcon = getWidgetAssetRegistry().loadTexture(L"res/editor/plus.png");
+				static pyr::Texture RemoveIcon = getWidgetAssetRegistry().loadTexture(L"res/editor/retirer.png");
+				static auto makeSelectable = [&](const char* label, int id, pf_Light* ref)
```
Comment (Akahara) [inline PyriteEditor/src/editor/views/Lights/widget_lights.h:27]: You musn't capture anything when using `static` before a lambda

File: PyriteEditor/src/editor/views/Lights/widget_lights.h [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+				ImGui::Checkbox("IsOn", &light->sourceLight->isOn);
+				ImGui::Separator();
+				ImGui::ColorEdit3("Ambiant", &light->sourceLight->ambiant.x);
+				ImGui::ColorEdit3("Diffuse", &light->sourceLight->diffuse.x);
+				ImGui::Separator();
+				switch (light->type)
+				{
+				case pf_Light::DIR:
+				{
+					pyr::DirectionalLight* sourceLight = static_cast<pyr::DirectionalLight*>(light->sourceLight);
```
Comment (Akahara) [inline PyriteEditor/src/editor/views/Lights/widget_lights.h:128]: That looks like a `variant` in disguise, but as-is is fine-ish enough

File: PyriteEditor/src/editor/views/Lights/widget_lights.h [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+		//		default:
+		//			break;
+		//		}
+		//
+		//		ImGui::EndChild();
+		//	}
+		//
+		//	ImGui::End();
+		//	return returnLight;
+		//}
```
Comment (Akahara) [inline PyriteEditor/src/editor/views/Lights/widget_lights.h:288]: Nice comment

File: PyriteEditor/src/editor/bridges/Lights/pf_Light.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+
+		// -- // 
+		std::vector<pf_Light> EditorFormat;
+		int selectedId = 0;
+		pf_Light* selectedLight = nullptr;
+
+	public:
+
+		void UpdateEditorFormat()
```
Comment (Akahara) [inline PyriteEditor/src/editor/bridges/Lights/pf_Light.h:42]: Not a fan of having this here, this class is half-data and half-logic

File: PyriteEditor/src/main.cpp [hunk 8]
```diff
@@ -15,14 +15,16 @@
...
 #include "utils/Debug.h"
 #include "engine/Engine.h"
 #include "engine/Device.h"
 #include "scene/SceneManager.h"
 #include "utils/StringUtils.h"
 #include "editor/ShaderReloader.h"
 
-int APIENTRY _tWinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPTSTR lpCmdLine, int nCmdShow)
+//int APIENTRY _tWinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPTSTR lpCmdLine, int nCmdShow)
+int main(const char** argv, int argc)
```
Comment (Akahara) [inline PyriteEditor/src/main.cpp:27]: Could be done nicelyerly someday

File: PyriteEditor/src/main.cpp [hunk 9]
```diff
@@ -43,9 +47,10 @@ int APIENTRY _tWinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPTSTR lpCm
     scenes.registerScene<pye::VoxelisationDemoScene>("Voxelisation demo");
     scenes.registerScene<pye::MaterialScene>("GGX Demo");
     scenes.registerScene<pye::CubemapBuilderScene>("IBL");
+    scenes.registerScene<pye::CornellBoxScene>("CornellBox");
 
     // Load the scene that is passed on the command line by default
-    scenes.transitionToScene(pyr::widestring2string(lpCmdLine));
+    //scenes.transitionToScene(pyr::widestring2string(lpCmdLine));
```
Comment (Akahara) [inline PyriteEditor/src/main.cpp:53]: Why not?

File: PyriteCore/src/world/Material.h [hunk 10]
```diff
@@ -37,9 +37,6 @@ class Material
 {
 public:
 
-
-   
-
 private:
```
[the remark this answers is not in the document]
Comment (JoshuaVHB) [reply PyriteCore/src/world/Material.h:40]: space. public

File: PyriteEditor/src/CornellBoxScene.h (hunk 1 above)
Comment (JoshuaVHB) [reply PyriteEditor/src/CornellBoxScene.h:139]: how about a PreRender() function

File: PyriteEditor/src/editor/views/Lights/widget_lights.h (hunk 5 above)
Comment (JoshuaVHB) [reply PyriteEditor/src/editor/views/Lights/widget_lights.h:128]: it is indeed

File: PyriteEditor/src/editor/bridges/Lights/pf_Light.h (hunk 7 above)
Comment (JoshuaVHB) [reply PyriteEditor/src/editor/bridges/Lights/pf_Light.h:42]: Not entirely sure i agree on this one. The goal is to have this half data / half logic, which is basically a model-view-view-model design

File: PyriteEditor/src/CornellBoxScene.h (hunk 2 above)
Comment (JoshuaVHB) [reply PyriteEditor/src/CornellBoxScene.h:151]: yeah but bothers me to do it so i'll do it later

**Pull request merged**